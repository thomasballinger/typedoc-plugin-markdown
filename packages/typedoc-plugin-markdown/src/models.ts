import { TypeDocOptionMap } from 'typedoc';

export interface TypedocPluginMarkdownOptions extends TypeDocOptionMap {
  hideBreadcrumbs: boolean;
  hideInPageTOC: boolean;
  hidePageTitle: boolean;
  hideMembersSymbol: boolean;
  entryDocument: string;
  indexTitle: string;
  namedAnchors: boolean;
  publicPath: string;
}

export type Collapse = 'object' | 'function' | 'all' | 'none';
