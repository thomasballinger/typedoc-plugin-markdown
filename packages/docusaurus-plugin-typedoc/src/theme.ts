import * as fs from 'fs';
import * as path from 'path';
import {
  BindOption,
  DeclarationReflection,
  PageEvent,
  ReflectionKind,
  Renderer,
  RendererEvent,
  UrlMapping,
} from 'typedoc';
import {
  FrontMatterVars,
  getKindPlural,
  getPageTitle,
  MarkdownTheme,
  prependYAML,
} from 'typedoc-plugin-markdown';
import { FrontMatter, SidebarOptions } from './types';

const CATEGORY_POSITION = {
  [ReflectionKind.Module]: 1,
  [ReflectionKind.Namespace]: 1,
  [ReflectionKind.Enum]: 2,
  [ReflectionKind.Class]: 3,
  [ReflectionKind.Interface]: 4,
  [ReflectionKind.TypeAlias]: 5,
  [ReflectionKind.Variable]: 6,
  [ReflectionKind.Function]: 7,
  [ReflectionKind.ObjectLiteral]: 8,
};

export class DocusaurusTheme extends MarkdownTheme {
  @BindOption('sidebar')
  sidebar!: SidebarOptions;

  @BindOption('readmeTitle')
  readmeTitle!: string;

  @BindOption('indexSlug')
  indexSlug!: string;

  @BindOption('frontmatter')
  frontmatter!: FrontMatter;

  constructor(renderer: Renderer) {
    super(renderer);

    this.listenTo(this.application.renderer, {
      [PageEvent.END]: this.onPageEnd,
      [RendererEvent.END]: this.onRendererEnd,
    });
  }

  onPageEnd(page: PageEvent<DeclarationReflection>) {
    if (page.contents) {
      page.contents = prependYAML(
        page.contents,
        this.getYamlItems(page) as FrontMatterVars,
      );
    }
  }

  onRendererEnd(renderer: RendererEvent) {
    if (this.sidebar.autoConfiguration) {
      writeCategoryYaml(
        renderer.outputDirectory,
        this.sidebar.categoryLabel,
        this.sidebar.position,
      );

      Object.keys(groupUrlsByKind(this.getUrls(renderer.project))).forEach(
        (group) => {
          const kind = parseInt(group);
          const mapping = this.mappings.find((mapping) =>
            mapping.kind.includes(kind),
          );
          if (mapping) {
            writeCategoryYaml(
              renderer.outputDirectory + '/' + mapping.directory,
              getKindPlural(kind),
              CATEGORY_POSITION[kind],
            );
          }
        },
      );
    }
  }

  getYamlItems(page: PageEvent<DeclarationReflection>): FrontMatter {
    const pageId = this.getId(page);
    const pageTitle = this.getTitle(page);
    const sidebarLabel = this.getSidebarLabel(page);
    const sidebarPosition = this.getSidebarPosition(page);
    let items: FrontMatter = {
      id: pageId,
      title: pageTitle,
    };
    if (page.url === this.entryDocument && this.indexSlug) {
      items = { ...items, slug: this.indexSlug };
    }

    if (this.sidebar.autoConfiguration) {
      if (sidebarLabel && sidebarLabel !== pageTitle) {
        items = { ...items, sidebar_label: sidebarLabel as string };
      }
      if (sidebarPosition) {
        items = { ...items, sidebar_position: parseFloat(sidebarPosition) };
      }
    }

    if (page.url === page.project.url && this.entryPoints.length > 1) {
      items = { ...items, hide_table_of_contents: true };
    }
    items = { ...items, custom_edit_url: null };
    if (this.frontmatter) {
      items = { ...items, ...this.frontmatter };
    }
    return {
      ...items,
    };
  }

  getSidebarLabel(page: PageEvent<DeclarationReflection>) {
    const indexLabel =
      this.sidebar.indexLabel ||
      (this.entryPoints.length > 1 ? 'Table of Contents' : 'Exports');

    if (page.url === this.entryDocument) {
      return page.url === page.project.url
        ? indexLabel
        : this.sidebar.readmeLabel;
    }

    if (page.url === this.getRenderContext().globalsFile) {
      return indexLabel;
    }

    return this.sidebar.fullNames ? page.model.getFullName() : page.model.name;
  }

  getSidebarPosition(page: PageEvent<DeclarationReflection>) {
    if (page.url === this.entryDocument) {
      return page.url === page.project.url ? '0.5' : '0';
    }
    if (page.url === this.getRenderContext().globalsFile) {
      return '0.5';
    }
    if (page.model.getFullName().split('.').length === 1) {
      return '0';
    }
    return null;
  }

  getId(page: PageEvent) {
    return path.basename(page.url, path.extname(page.url));
  }

  getTitle(page: PageEvent) {
    const readmeTitle = this.readmeTitle || page.project.name;
    if (page.url === this.entryDocument && page.url !== page.project.url) {
      return readmeTitle;
    }
    return getPageTitle(page);
  }

  get mappings() {
    return super.mappings.map((mapping) => {
      if (mapping.kind.includes(ReflectionKind.Namespace)) {
        return {
          ...mapping,
          directory: 'namespaces',
        };
      }
      return mapping;
    });
  }
}

const writeCategoryYaml = (
  categoryPath: string,
  label: string,
  position: number | null,
) => {
  const yaml: string[] = [`label: "${label}"`];
  if (position !== null) {
    yaml.push(`position: ${position}`);
  }
  if (fs.existsSync(categoryPath)) {
    fs.writeFileSync(categoryPath + '/_category_.yml', yaml.join('\n'));
  }
};

const groupUrlsByKind = (urls: UrlMapping[]) => {
  return urls.reduce(
    (r, v, i, a, k = v.model.kind) => ((r[k] || (r[k] = [])).push(v), r),
    {},
  );
};
