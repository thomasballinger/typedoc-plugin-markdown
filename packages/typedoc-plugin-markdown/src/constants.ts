import { ReflectionKind } from 'typedoc';

export const URL_PREFIX = /^(http|ftp)s?:\/\//;

export const PLURALS = {
  [ReflectionKind.Class]: 'Classes',
  [ReflectionKind.Property]: 'Properties',
  [ReflectionKind.Enum]: 'Enumerations',
  [ReflectionKind.EnumMember]: 'Enumeration members',
  [ReflectionKind.TypeAlias]: 'Type aliases',
};
