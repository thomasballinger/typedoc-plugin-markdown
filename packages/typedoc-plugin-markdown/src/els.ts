export const heading = (level: number, text: string) =>
  `${[...Array(level)].map(() => '#').join('')} ${text}`;

export const link = (label: string, url: string | null) =>
  url ? `[${label}](${url})` : '';

export const bold = (text: string) => `**${text}**`;

export const backTicks = (text: string) => `\`${text}\``;

export const unorderedList = <T>(items: T[]) =>
  items.map((item) => `- ${item}`).join('\n');

export const horizontalRule = () => '---';

export const table = (headers: string[], rows: string[][]) =>
  `\n| ${headers.join(' | ')} |\n| ${headers
    .map(() => ':------')
    .join(' | ')} |\n${rows.map((row) => `| ${row.join(' | ')} \n`).join('')}`;
