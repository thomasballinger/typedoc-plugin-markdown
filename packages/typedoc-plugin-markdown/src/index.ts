import { Application, Options, OptionsReader, ParameterType } from 'typedoc';

import { MarkdownTheme } from './theme';
import { MarkdownThemeRenderContext } from './theme-context';

export function load(app: Application) {
  app.renderer.defineTheme('markdown', MarkdownTheme);
  app.options.addReader(
    new (class implements OptionsReader {
      priority = 1000;
      name = 'markdown-theme-reader';
      read(container: Options) {
        if (container.getValue('theme') === 'default') {
          container.setValue('theme', 'markdown');
        }
      }
    })(),
  );

  app.options.addDeclaration({
    help: '[Markdown Plugin] Do not render page title.',
    name: 'hidePageTitle',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Do not render breadcrumbs in template.',
    name: 'hideBreadcrumbs',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Specifies the base path that all links to be served from. If omitted all urls will be relative.',
    name: 'publicPath',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Use HTML named anchors as fragment identifiers for engines that do not automatically assign header ids. Should be set for Bitbucket Server docs.',
    name: 'namedAnchors',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Output all reflections into seperate output files.',
    name: 'allReflectionsHaveOwnDocument',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Separator used to format filenames.',
    name: 'filenameSeparator',
    type: ParameterType.String,
    defaultValue: '.',
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] The file name of the entry document.',
    name: 'entryDocument',
    type: ParameterType.String,
    defaultValue: 'README.md',
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Do not render in-page table of contents items.',
    name: 'hideInPageTOC',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Customise the index page title.',
    name: 'indexTitle',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Do not add special symbols for class members.',
    name: 'hideMembersSymbol',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Preserve anchor casing when generating links.',
    name: 'preserveAnchorCasing',
    type: ParameterType.Boolean,
    defaultValue: false,
  });
}

export { MarkdownTheme, MarkdownThemeRenderContext };
