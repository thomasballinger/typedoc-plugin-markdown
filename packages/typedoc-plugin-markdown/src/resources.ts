import { 
 ArrayType,
 ConditionalType,
 ContainerReflection,
 Comment,
 CommentDisplayPart,
 DeclarationHierarchy,
 DeclarationReflection,
 InferredType,
 IntersectionType,
 IntrinsicType,
 IndexedAccessType,
 LiteralType,
 PageEvent,
 ParameterReflection,
 ProjectReflection,
 QueryType,
 ReflectionGroup,
 ReferenceReflection,
 ReferenceType,
 ReflectionType,
 Reflection,
 SignatureReflection,
 SomeType,
 TupleType,
 TypeOperatorType,
 TypeParameterReflection,
 UnionType,
 UnknownType } from 'typedoc';

import { MarkdownThemeRenderContext } from './theme-context';
import { Collapse } from './models';
import { project } from './templates/project';
import { readme } from './templates/readm3';
import { reflection } from './templates/reflection';
import { breadcrumbs } from './partials/breadcrumbs';
import { commentParts } from './partials/comment.parts';
import { comment } from './partials/comment';
import { hierarchy } from './partials/hierarchy';
import { declarationTitle } from './partials/member.declaration.title';
import { declaration } from './partials/member.declaration';
import { sources } from './partials/member.sources';
import { member } from './partials/member';
import { memberReference } from './partials/memberReference';
import { members } from './partials/members';
import { membersGroup } from './partials/membersGroup';
import { memberSignature } from './partials/memberSignature';
import { parameterTable } from './partials/parameterTable';
import { propertyTable } from './partials/propertyTable';
import { reflectionTitle } from './partials/reflectionTitle';
import { toc } from './partials/toc';
import { arrayType } from './partials/type.array';
import { conditionalType } from './partials/type.conditional';
import { declarationType } from './partials/type.declaration';
import { functionType } from './partials/type.function';
import { indexAccessType } from './partials/type.indexAccess';
import { inferredType } from './partials/type.inferred';
import { intersectionType } from './partials/type.intersection';
import { intrinsicType } from './partials/type.intrinsic';
import { literalType } from './partials/type.literal';
import { queryType } from './partials/type.query';
import { referenceType } from './partials/type.reference';
import { reflectionType } from './partials/type.reflection';
import { someType } from './partials/type.some';
import { tupleType } from './partials/type.tuple';
import { typeOperatorType } from './partials/type.typeOperator';
import { unionType } from './partials/type.union';
import { unknownType } from './partials/type.unknown';
import { typeParameters } from './partials/typeParameters';
import { typeParameterTable } from './partials/typeParameterTable';

function bind<F, L extends any[], R>(fn: (f: F, ...a: L) => R, first: F) {
  return (...r: L) => fn(first, ...r);
}
export type Templates = {
  project: (page: PageEvent<ProjectReflection>) => string;
  readme: (page: PageEvent<ProjectReflection>) => string;
  reflection: (page: PageEvent<DeclarationReflection>) => string;
};

export type Partials = {
  breadcrumbs: (page: PageEvent<ProjectReflection | DeclarationReflection>) => string;
  commentParts: (parts: CommentDisplayPart[]) => string;
  comment: (comment: Comment) => string;
  hierarchy: (declarationHierarchy: DeclarationHierarchy) => string;
  declarationTitle: (reflection: DeclarationReflection | ParameterReflection) => string;
  declaration: (reflection: DeclarationReflection) => string;
  sources: (reflection: DeclarationReflection | SignatureReflection) => string;
  member: (props: DeclarationReflection) => string;
  memberReference: (props: ReferenceReflection) => string;
  members: (props: ContainerReflection) => string;
  membersGroup: (group: ReflectionGroup) => string;
  memberSignature: (props: SignatureReflection, nested?: boolean, accessor?: string |  undefined) => string;
  parameterTable: (parameters: ParameterReflection[]) => string;
  propertyTable: (props: DeclarationReflection[]) => string;
  reflectionTitle: (declarationReflection: DeclarationReflection, shouldEscape?: boolean) => string;
  toc: (props: ProjectReflection | DeclarationReflection) => string;
  arrayType: (arrayType: ArrayType, emphasis: boolean) => string;
  conditionalType: (conditionalType: ConditionalType) => string;
  declarationType: (declarationReflection: DeclarationReflection, collapse?: Collapse) => string;
  functionType: (modelSignatures: SignatureReflection[]) => string;
  indexAccessType: (model: IndexedAccessType) => string;
  inferredType: (model: InferredType) => string;
  intersectionType: (model: IntersectionType) => string;
  intrinsicType: (model: IntrinsicType, emphasis: boolean) => string;
  literalType: (literalType: LiteralType) => string;
  queryType: (queryType: QueryType) => string;
  referenceType: (referenceType: ReferenceType, emphasis: boolean) => string;
  reflectionType: (reflectionType: ReflectionType, collapse: Collapse) => string;
  someType: (someType: SomeType, collapse?: Collapse, emphasis?: boolean) => string;
  tupleType: (tupleType: TupleType) => string;
  typeOperatorType: (model: TypeOperatorType) => string;
  unionType: (unionType: UnionType, emphasis: boolean) => string;
  unknownType: (model: UnknownType) => string;
  typeParameters: (typeParameters: TypeParameterReflection[]) => string;
  typeParameterTable: (props: TypeParameterReflection[]) => string;
};

export const templates = (context: MarkdownThemeRenderContext): Templates => ({
  project: bind(project, context),
  readme: bind(readme, context),
  reflection: bind(reflection, context),
});

export const partials = (context: MarkdownThemeRenderContext): Partials => ({
  breadcrumbs: bind(breadcrumbs, context),
  commentParts: bind(commentParts, context),
  comment: bind(comment, context),
  hierarchy: bind(hierarchy, context),
  declarationTitle: bind(declarationTitle, context),
  declaration: bind(declaration, context),
  sources: bind(sources, context),
  member: bind(member, context),
  memberReference: bind(memberReference, context),
  members: bind(members, context),
  membersGroup: bind(membersGroup, context),
  memberSignature: bind(memberSignature, context),
  parameterTable: bind(parameterTable, context),
  propertyTable: bind(propertyTable, context),
  reflectionTitle: bind(reflectionTitle, context),
  toc: bind(toc, context),
  arrayType: bind(arrayType, context),
  conditionalType: bind(conditionalType, context),
  declarationType: bind(declarationType, context),
  functionType: bind(functionType, context),
  indexAccessType: bind(indexAccessType, context),
  inferredType: bind(inferredType, context),
  intersectionType: bind(intersectionType, context),
  intrinsicType: bind(intrinsicType, context),
  literalType: bind(literalType, context),
  queryType: bind(queryType, context),
  referenceType: bind(referenceType, context),
  reflectionType: bind(reflectionType, context),
  someType: bind(someType, context),
  tupleType: bind(tupleType, context),
  typeOperatorType: bind(typeOperatorType, context),
  unionType: bind(unionType, context),
  unknownType: bind(unknownType, context),
  typeParameters: bind(typeParameters, context),
  typeParameterTable: bind(typeParameterTable, context),
});