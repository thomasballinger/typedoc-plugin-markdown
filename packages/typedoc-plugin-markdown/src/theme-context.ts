import * as path from 'path';
import { Options, ReferenceType, Reflection } from 'typedoc';
import { URL_PREFIX } from './constants';
import { TypedocPluginMarkdownOptions } from './models';
import { partials, templates } from './resources';
import { MarkdownTheme } from './theme';

export class MarkdownThemeRenderContext {
  private _activeLocation: string;
  // private _project: ProjectReflection;
  private _globalsFile = 'modules.md';

  /*set project(project: ProjectReflection) {
    this._project = project;
  }

  get project() {
    return this._project;
  }
*/
  set activeLocation(activeLocation: string) {
    this._activeLocation = activeLocation;
  }

  get activeLocation() {
    return this._activeLocation;
  }

  set globalsFile(globalsFile: string) {
    this._globalsFile = globalsFile;
  }

  get globalsFile() {
    return this._globalsFile;
  }

  constructor(private theme: MarkdownTheme, public options: Options) {}

  getOption<K extends keyof TypedocPluginMarkdownOptions>(name: K) {
    return this.options.getValue(name) as TypedocPluginMarkdownOptions[K];
  }

  attemptExternalResolution = (type: ReferenceType) => {
    return this.theme.owner.attemptExternalResolution(type);
  };

  urlTo = (reflection: Reflection) => this.relativeURL(reflection.url);

  relativeURL(url: string | undefined) {
    if (!url) {
      return null;
    }
    if (URL_PREFIX.test(url)) {
      return url;
    } else {
      const relative = path.relative(
        path.dirname(this.activeLocation),
        path.dirname(url),
      );
      return path.join(relative, path.basename(url)).replace(/\\/g, '/');
    }
  }

  templates = templates(this);
  partials = partials(this);
}
