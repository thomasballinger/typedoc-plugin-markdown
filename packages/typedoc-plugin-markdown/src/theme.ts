import {
  BindOption,
  DeclarationReflection,
  PageEvent,
  ProjectReflection,
  Reflection,
  ReflectionKind,
  Renderer,
  RendererEvent,
  Theme,
  UrlMapping,
} from 'typedoc';
import { URL_PREFIX } from './constants';
import { MarkdownThemeRenderContext } from './theme-context';
import { formatContents } from './utils';

export class MarkdownTheme extends Theme {
  @BindOption('allReflectionsHaveOwnDocument')
  allReflectionsHaveOwnDocument!: boolean;
  @BindOption('entryDocument') entryDocument!: string;
  @BindOption('entryPoints') entryPoints!: string[];
  @BindOption('filenameSeparator') filenameSeparator!: string;
  @BindOption('readme') readme!: string;
  @BindOption('preserveAnchorCasing') preserveAnchorCasing!: boolean;

  private _renderContext?: MarkdownThemeRenderContext;
  private anchorMap: Record<string, string[]> = {};

  constructor(renderer: Renderer) {
    super(renderer);

    this.listenTo(this.owner, {
      [RendererEvent.BEGIN]: this.onBeginRenderer,
      [PageEvent.BEGIN]: this.onBeginPage,
    });
  }

  getRenderContext() {
    if (!this._renderContext) {
      this._renderContext = new MarkdownThemeRenderContext(
        this,
        this.application.options,
      );
    }
    return this._renderContext;
  }

  readmeTemplate = (pageEvent: PageEvent<ProjectReflection>) => {
    return this.getRenderContext().templates.readme(pageEvent);
  };

  projectTemplate = (pageEvent: PageEvent<ProjectReflection>) => {
    return this.getRenderContext().templates.project(pageEvent);
  };

  reflectionTemplate = (pageEvent: PageEvent<DeclarationReflection>) => {
    return this.getRenderContext().templates.reflection(pageEvent);
  };

  render(page: PageEvent<Reflection>): string {
    return formatContents(page.template(page) as string);
  }

  getUrls(project: ProjectReflection) {
    const urls: UrlMapping[] = [];
    const noReadmeFile = this.readme.endsWith('none');
    if (noReadmeFile) {
      project.url = this.entryDocument;
      urls.push(
        new UrlMapping(this.entryDocument, project, this.projectTemplate),
      );
    } else {
      project.url = this.getRenderContext().globalsFile;
      urls.push(
        new UrlMapping(
          this.getRenderContext().globalsFile,
          project,
          this.projectTemplate,
        ),
      );
      urls.push(
        new UrlMapping(this.entryDocument, project, this.readmeTemplate),
      );
    }
    project.children?.forEach((child: Reflection) => {
      if (child instanceof DeclarationReflection) {
        this.buildUrls(child as DeclarationReflection, urls);
      }
    });
    return urls;
  }

  buildUrls(
    reflection: DeclarationReflection,
    urls: UrlMapping[],
  ): UrlMapping[] {
    const mapping = this.mappings.find((mapping) =>
      reflection.kindOf(mapping.kind),
    );
    if (mapping) {
      if (!reflection.url || !URL_PREFIX.test(reflection.url)) {
        const url = this.toUrl(mapping, reflection);
        urls.push(new UrlMapping(url, reflection, mapping.template));
        reflection.url = url;
        reflection.hasOwnDocument = true;
      }

      for (const child of reflection.children || []) {
        if (mapping.isLeaf) {
          this.applyAnchorUrl(child, reflection);
        } else {
          this.buildUrls(child, urls);
        }
      }
    } else if (reflection.parent) {
      this.applyAnchorUrl(reflection, reflection.parent, true);
    }
    return urls;
  }

  toUrl(mapping: any, reflection: DeclarationReflection) {
    return mapping.directory + '/' + this.getUrl(reflection) + '.md';
  }

  getUrl(reflection: Reflection, relative?: Reflection): string {
    let url = reflection.getAlias();

    if (
      reflection.parent &&
      reflection.parent !== relative &&
      !(reflection.parent instanceof ProjectReflection)
    ) {
      url =
        this.getUrl(reflection.parent, relative) + this.filenameSeparator + url;
    }

    return url.replace(/^_/, '');
  }

  applyAnchorUrl(
    reflection: Reflection,
    container: Reflection,
    isSymbol = false,
  ) {
    if (
      container.url &&
      (!reflection.url || !URL_PREFIX.test(reflection.url))
    ) {
      const reflectionId = this.preserveAnchorCasing
        ? reflection.name
        : reflection.name.toLowerCase();

      if (isSymbol) {
        this.anchorMap[container.url]
          ? this.anchorMap[container.url].push(reflectionId)
          : (this.anchorMap[container.url] = [reflectionId]);
      }

      const count = this.anchorMap[container.url]?.filter(
        (id) => id === reflectionId,
      )?.length;

      const anchor = this.toAnchorRef(
        reflectionId + (count > 1 ? '-' + (count - 1).toString() : ''),
      );

      reflection.url = container.url + '#' + anchor;
      reflection.anchor = anchor;
      reflection.hasOwnDocument = false;
    }
    reflection.traverse((child) => {
      if (child instanceof DeclarationReflection) {
        this.applyAnchorUrl(child, container);
      }
    });
  }

  toAnchorRef(reflectionId: string) {
    return reflectionId;
  }

  get mappings() {
    return [
      {
        kind: [ReflectionKind.Module],
        isLeaf: false,
        directory: 'modules',
        template: this.reflectionTemplate,
      },
      {
        kind: [ReflectionKind.Namespace],
        isLeaf: false,
        directory: 'modules',
        template: this.reflectionTemplate,
      },
      {
        kind: [ReflectionKind.Enum],
        isLeaf: false,
        directory: 'enums',
        template: this.reflectionTemplate,
      },
      {
        kind: [ReflectionKind.Class],
        isLeaf: false,
        directory: 'classes',
        template: this.reflectionTemplate,
      },
      {
        kind: [ReflectionKind.Interface],
        isLeaf: false,
        directory: 'interfaces',
        template: this.reflectionTemplate,
      },
      ...(this.allReflectionsHaveOwnDocument
        ? [
            {
              kind: [ReflectionKind.TypeAlias],
              isLeaf: true,
              directory: 'types',
              template: this.reflectionTemplate,
            },
            {
              kind: [ReflectionKind.Variable],
              isLeaf: true,
              directory: 'variables',
              template: this.reflectionTemplate,
            },
            {
              kind: [ReflectionKind.Function],
              isLeaf: true,
              directory: 'functions',
              template: this.reflectionTemplate,
            },
          ]
        : []),
    ];
  }

  protected onBeginRenderer(event: RendererEvent) {
    //this.getRenderContext().project = event.project;
  }

  protected onBeginPage(page: PageEvent) {
    this.getRenderContext().activeLocation = page.url;
  }
}
