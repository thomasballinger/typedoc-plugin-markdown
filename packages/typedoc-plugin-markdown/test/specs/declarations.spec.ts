import { ProjectReflection } from 'typedoc';
import { MarkdownThemeRenderContext } from '../../src/theme-context';

describe(`Declarations:`, () => {
  let project: ProjectReflection;
  let context: MarkdownThemeRenderContext;

  beforeAll(async () => {
    ({ project, context } = await global.bootstrap(['declarations.ts'], {
      stubPartials: ['sources'],
    }));
  });

  test(`should compile a const with default value`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('stringConstWithDefaultValue') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile a let with default value`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('stringLetWithDefaultValue') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile an undefined declaration`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('undefinedNumberDeclaration') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal declaration`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('objectLiteralDeclaration') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal cast as a const`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('objectLiteralAsConstDeclaration') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type literal declaration`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('typeLiteralDeclaration') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile declaration with double underscores in name and value`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('__DOUBLE_UNDERSCORES_DECLARATION__') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile any function type`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('AnyFunctionType') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile function declaration`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('functionDeclaration') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile callable declaration`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('callableDeclaration') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile indexable declaration`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('indexableDeclaration') as any,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile enum declaration`, () => {
    expect(
      context.partials.declaration(
        (project.getChildByName('EnumDeclarations') as any).children[0],
      ),
    ).toMatchSnapshot();
  });

  test(`should compile enum declaration with defaults`, () => {
    expect(
      context.partials.declaration(
        (project.getChildByName('EnumDeclarationsWithDefaults') as any)
          .children[0],
      ),
    ).toMatchSnapshot();
  });

  test(`should compile declaration with accessors`, () => {
    expect(
      context.partials.declaration(
        project.getChildByName('getterAndSetter') as any,
      ),
    ).toMatchSnapshot();
  });
});
