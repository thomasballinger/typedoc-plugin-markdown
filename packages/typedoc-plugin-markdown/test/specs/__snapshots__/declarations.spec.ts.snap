// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Declarations: should compile a const with default value 1`] = `
" \`Const\` **stringConstWithDefaultValue**: \\"hello\\"

[partial: sources]"
`;

exports[`Declarations: should compile a let with default value 1`] = `
" **stringLetWithDefaultValue**: \`string\` = \`'hello'\`

[partial: sources]"
`;

exports[`Declarations: should compile an undefined declaration 1`] = `
" **undefinedNumberDeclaration**: \`number\`

[partial: sources]"
`;

exports[`Declarations: should compile any function type 1`] = `
" **AnyFunctionType**<\`A\`\\\\>: (...\`input\`: \`any\`[]) => \`A\`

#### Type parameters

| Name | Type |
| :------ | :------ |
| \`A\` | \`any\` |

#### Type declaration

(...\`input\`: \`any\`[]): \`A\`

##### Parameters]


| Name | Type |
| :------ | :------ |
| \`...input\` | \`any\`[] |


##### Returns

\`A\`

[partial: sources]"
`;

exports[`Declarations: should compile callable declaration 1`] = `
" **callableDeclaration**: \`Object\`

#### Call signature

(\`someArg\`: \`number\`): \`boolean\`

##### Parameters


| Name | Type |
| :------ | :------ |
| \`someArg\` | \`number\` |


##### Returns

\`boolean\`

#### Type declaration


| Name | Type |
| :------ | :------ |
| \`arg1\` | \`string\` |
| \`arg2\` | \`number\` |


[partial: sources]"
`;

exports[`Declarations: should compile declaration with accessors 1`] = `
" **getterAndSetter**: \`Object\`

#### Type declaration


| Name | Type |
| :------ | :------ |
| \`get getMe()\` | [\`Promise\`]( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise )<\`string\`\\\\> |
| \`set setMe(x:string)\` | \`void\` |


[partial: sources]"
`;

exports[`Declarations: should compile declaration with double underscores in name and value 1`] = `
" \`Const\` **\\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_**: typeof [\`__DOUBLE_UNDERSCORES_DECLARATION__\`](../modules.md#__double_underscores_declaration__)

[partial: sources]"
`;

exports[`Declarations: should compile enum declaration 1`] = `
" **Down** = 1

[partial: sources]"
`;

exports[`Declarations: should compile enum declaration with defaults 1`] = `
" **East** = \\"East\\"

[partial: sources]"
`;

exports[`Declarations: should compile function declaration 1`] = `
" **functionDeclaration**: 

[partial: sources]"
`;

exports[`Declarations: should compile indexable declaration 1`] = `
" **indexableDeclaration**: \`Object\`

context.indexSignaturePartial(typeDeclaration.indexSignature)

#### Type declaration


| Name | Type |
| :------ | :------ |
| \`arg1\` | \`string\` |


[partial: sources]"
`;

exports[`Declarations: should compile object literal cast as a const 1`] = `
" \`Const\` **objectLiteralAsConstDeclaration**: \`Object\`

Comments

**\`Param\`**

Comment for object.

**\`Param\`**

Comment for Prop1.

**\`Param\`**

Comment for Prop2.

#### Type declaration


| Name | Type |
| :------ | :------ |
| \`Prop1\` | \\"Prop1\\" |
| \`Prop2\` | \\"Prop2\\" |
| \`Prop3\` | \\"Prop3\\" |


[partial: sources]"
`;

exports[`Declarations: should compile object literal declaration 1`] = `
" \`Const\` **objectLiteralDeclaration**: \`Object\`

**\`Param\`**

description for valueX

**\`Param\`**

description for valueZ

**\`Param\`**

description for valueY

#### Type declaration


| Name | Type | Description |
| :------ | :------ | :------ |
| \`valueA\` | \`number\` | Comment for valueA |
| \`valueB\` | \`boolean\` | - |
| \`valueC\` | {} | - |
| \`valueX\` | { \`valueA\`: \`number\`[] ; \`valueZ\`: \`string\` = 'foo' } | Comment for valueX |
| \`valueX.valueA\` | \`number\`[] | - |
| \`valueX.valueZ\` | \`string\` | - |
| \`valueY\` | () => \`string\` | Comment for value Y |
| \`valueZ\` | \`string\` | Comment for valueZ |


[partial: sources]"
`;

exports[`Declarations: should compile type literal declaration 1`] = `
" **typeLiteralDeclaration**: \`Object\`

#### Type declaration


| Name | Type | Description |
| :------ | :------ | :------ |
| \`valueA?\` | \`number\` | Comments for valueA |
| \`valueB?\` | \`boolean\` | - |
| \`valueC\` | { \`[dataId: string]\`: \\"ok\\" \\\\| \\"ko\\";  } | - |
| \`valueX\` | { \`valueA\`: \`number\`[] ; \`valueY\`: (\`z\`: \`string\`) => { \`a\`: \`string\` ; \`b\`: \`string\`  } ; \`valueZ\`: \`string\`  } | Comment for valueX |
| \`valueX.valueA\` | \`number\`[] | - |
| \`valueX.valueY\` | (\`z\`: \`string\`) => { \`a\`: \`string\` ; \`b\`: \`string\`  } | - |
| \`valueX.valueZ\` | \`string\` | Nested comment for valueZ |
| \`valueY\` | () => \`string\` | - |
| \`valueZ\` | \`string\` | Comment for valueZ |


[partial: sources]"
`;
